tools/src/main/scala/scala/scalanative/optimizer/pass/SRemOverflow.scala

```diff
@@ -0,0 +1,68 @@
+package scala.scalanative.optimizer.pass
+
+import scala.scalanative.nir
+import scala.scalanative.nir._
+import scala.scalanative.optimizer.analysis.ClassHierarchy.Top
+import scala.scalanative.optimizer.{Pass, PassCompanion}
+import scala.scalanative.tools.Config
+
+/**
+ * Detects taking remainder for division by -1 and replaces it by division by 1 which can't overflow.
+ *
+ *
+ * We implement '%' (remainder) with LLVM's 'srem' and it can overflow for cases:
+ * Int.MinValue % -1
+ * Long.MinValue % -1
+ * E.g. On x86_64 'srem' might get translated to 'idiv' which computes both quotient and remainder at once
+ * and quotient can overflow.
+ */
+class SRemOverflow(implicit top: Top) extends Pass {
+
+  override def onInsts(insts: Seq[Inst]): Seq[Inst] = {
+    val buf = new nir.Buffer
+    import buf._
+
+    def matchSrem(IntType: Type,
+                  makeVal: Int => Val): PartialFunction[Inst, Unit] = {
+
+      case Inst.Let(name, sremBin @ Op.Bin(Bin.Srem, IntType, _, divisor)) =>
+        val safeDivisor         = Val.Local(fresh(), IntType)
+        val thenL, elseL, contL = fresh()
+
+        val isPossibleOverflow = let(
+          Op.Comp(Comp.Ieq, IntType, divisor, makeVal(-1)))
+        branch(isPossibleOverflow, Next(thenL), Next(elseL))
```

Comment:
@densh Are there conditional moves in your IR?

If yest, this should definitely be a conditional move instead of a conditional branch.

Reply:
LLVM doesn't expose conditional moves explicitly. AFAIK instruction selection does pick cmov if it feels like it.

Reply:
In the release mode we are indeed getting a conditional move:

```
  def foo(value: Long, divisor: Long): Long = {
    value % divisor
  }
```

```
	.type	"Test$::foo_i64_i64_i64",@function
"Test$::foo_i64_i64_i64":               # @"Test$::foo_i64_i64_i64"
# BB#0:                                 # %_4.0
	cmp	rdx, -1
	mov	ecx, 1
	cmovne	rcx, rdx
	mov	rax, rsi
	cqo
	idiv	rcx
	mov	rax, rdx
	ret
```